src/js/select2/data/select.js

```diff
@@ -114,7 +114,12 @@ define([
     });
 
     container.on('unselect', function (params) {
-      self.unselect(params.data);
+      container.$element.find('option').each(function () {
+        if ($(this).val() == params.data.id) {
```

Comment:
Note: this is somewhat intentionally not a complete-strict equality test (in other words: it is expected that this is `==` instead of `===`)

This follows a similar equality pattern check in [`array.js` here](https://github.com/select2/select2/blob/5c38f186c98d793c7d396442986191ca05634d0a/src/js/select2/data/array.js#L47).

If both option-value and data-id are guaranteed to be of type `string`, then this could be upgraded to a strict equality test.